ruby/rspec/style_guide.md

```diff
@@ -0,0 +1,645 @@
+```
+
+### Mocks/Stubs/Doubles
+
+Use mocks and stubs with caution. While they help to improve the performance
+of the test suite, you can mock/stub yourself into a false-positive state very
+easily. When resorting to mocking and stubbing, only mock against a small,
+stable, obvious (or documented) API, so stubs are likely to represent reality
+after future refactoring.
+
+This generally means you should use them with more isolated/behavioral
+tests rather than with integration tests.
+
+```ruby
+# double an object
+article = double('article')
+
+# stubbing a method
+allow(Article).to receive(:find).with(5).and_return(article)
+```
+
+*NOTE*: if you stub a method that could give a false-positive test result, you
+have gone too far. See below:
+
+#### Bad Example
+
+```ruby
+subject { double('article') }
+
+describe '#summary' do
+  context 'when summary is not present' do
+    # This stubbing of the #nil? method, makes the test pass, but
+    # you are no longer testing the functionality of the code,
+    # you are testing the functionality of the test suite.
+    # This test would pass if there was not a single line of code
+    # written for the Article class.
+    it 'returns nil' do
+      summary = double('summary')
+      allow(subject).to receive(:summary).and_return(summary)
+      allow(summary).to receive(:nil?).and_return(true)
+      expect(subject.summary).to be_nil
+    end
+  end
+end
+```
+
+#### Good Example
+
+```ruby
+subject { double('article') }
+
+describe '#summary' do
+  context 'when summary is not present' do
+    # This is no longer stubbing all of the functionality, and will
+    # actually test the objects handling of the methods return value.
+    it 'returns nil' do
+      allow(subject).to receive(:summary).and_return(nil)
+      expect(subject.summary).to be_nil
+    end
+  end
+end
+```
+
+### Dealing with Time
+
+Always use [Timecop](https://github.com/travisjeffery/timecop) instead of
+stubbing anything on Time or Date.
+
+#### Bad Example
+
+```ruby
+it 'offsets the time 2 days into the future' do
+  current_time = Time.now
+  allow(Time).to receive(:now).and_return(current_time)
+  expect(subject.get_offset_time).to be_the_same_time_as (current_time + 2.days)
+end
+```
+
+#### Good Example
+
+```ruby
+it 'offsets the time 2 days into the future' do
+  Timecop.freeze(Time.now) do
+    expect(subject.get_offset_time).to eq 2.days.from_now
+  end
+end
+```
+
+
+### `let` blocks
+
+Use `let` blocks instead of `before(:each)` blocks to create data for the spec
+examples. `let` blocks get lazily evaluated. It also removes the instance
+variables from the test suite (which don't look as nice as local variables).
+
+These should primarily be used when you have duplication among a number of
+`it` blocks within a `context` but not all of them. Be careful with overuse of
+`let` as it makes the test suite much more difficult to read.
+
+```ruby
+# use this:
+let(:article) { FactoryGirl.create(:article) }
```

Comment:
We should be consistent with using `FactoryGirl.create` or `create`. I prefer the second.